HCore-Web/Providers/Impl/DownloadProcessingProxyUrlProviderImpl.cs

```diff
@@ -0,0 +1,197 @@
+    ///
+    /// <remarks>More processing is possible, of course.</remarks>
+    /// </summary>
+    public class DownloadProcessingProxyUrlProviderImpl : IDownloadProcessingProxyUrlProvider
+    {
+        public static readonly string ProcessingProxyBaseUrlConfig = "DownloadProxy:BaseUrl";
+        public static readonly string UrlQueryKeyName = "data";
+
+        private readonly ILogger<DownloadProcessingProxyUrlProviderImpl> _logger;
+        private readonly IDataProtector _protector;
+        private readonly IServiceProvider _serviceProvider;
+
+        public DownloadProcessingProxyUrlProviderImpl(
+            IDataProtectionProvider protectionProvider,
+            IServiceProvider serviceProvider,
+            ILogger<DownloadProcessingProxyUrlProviderImpl> logger
+        )
+        {
+            if (protectionProvider == null)
+            {
+                throw new ArgumentNullException(nameof(protectionProvider));
+            }
+
+            _protector = protectionProvider.CreateProtector(nameof(DownloadProcessingProxyUrlProviderImpl));
+            _serviceProvider = serviceProvider;
+            _logger = logger;
+        }
+
+
+        public virtual Task<Uri> CreateProxyUrlAsync(Uri downloadSourceUri, string fileName, string proxyBaseUrl, string downloadMimeType = null)
+        {
+            if (downloadSourceUri == null)
+            {
+                throw new ArgumentNullException(nameof(downloadSourceUri));
+            }
+
+            if (!downloadSourceUri.IsAbsoluteUri)
+            {
+                throw new ArgumentException("Download source URI is not absolute URI!");
+            }
+
+            if (string.IsNullOrEmpty(proxyBaseUrl))
+            {
+                throw new ArgumentException("No proxy base URL has been provided!");
+            }
+
+            // convert the data to JSON string
+            string uriQueryPayload = ConvertParametersToQueryData(CalculateHashFromParameters(downloadSourceUri, fileName, downloadMimeType));
+
+            UriBuilder proxyUri = new UriBuilder(proxyBaseUrl);
+            if (proxyUri.Query != null && proxyUri.Query.Length > 1)
+            {
+                proxyUri.Query = proxyUri.Query.Substring(1) + "&" + uriQueryPayload;
+            }
+            else
+            {
+                proxyUri.Query = uriQueryPayload;
+            }
+
+            proxyUri.Query = proxyUri.Query.Substring(1) + "&fileName=" + Uri.EscapeDataString(fileName)
+                + (downloadMimeType != null ? "&mime=" + Uri.EscapeDataString(downloadMimeType) : "")
+                + "&u=" + Uri.EscapeDataString(downloadSourceUri.ToString())
+            ;
+            return Task.FromResult(proxyUri.Uri);
+        }
+
+        public virtual async Task<IDownloadFileData> GetFileDataAsync(HttpRequest request, Stream inputData)
+        {
+            Uri downloadUri = new Uri(request.Query["u"]);
```

Comment:
I am not sure but what happens if the key does not exist in the dictionary?

Reply:
No exception is thrown but `StringValues.Empty` is returned. See [`IQueryCollection`](https://docs.microsoft.com/ru-ru/dotnet/api/microsoft.aspnetcore.http.iquerycollection.item?view=aspnetcore-3.0#Microsoft_AspNetCore_Http_IQueryCollection_Item_System_String_), which is the type of `Query` in class [`HttpRequest`](https://docs.microsoft.com/ru-ru/dotnet/api/microsoft.aspnetcore.http.httprequest.query?view=aspnetcore-3.0#Microsoft_AspNetCore_Http_HttpRequest_Query).

Since function `CreateProxyUrlAsync` checks whether original source download URL is empty and throws an exception, the URL may not be empty initially. So a valid URL will be used with the hash calculation. In case the parameter is empty at this place, the hash will not match and the call will fail with exception `UnauthorizedAccessException`.